aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -515,6 +515,369 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
   return result;
 }
 
+// helper methods for matrix_exp
+namespace {
+
+template <typename scalar_t, int ROW, int COL>
+using array2d = std::array<std::array<scalar_t, COL>, ROW>;
+
+// we consider 6 Taylor expansions of degree
+// 1, 2, 4, 8, 12, 18
+constexpr int total_n_degs = 6;
+
+// 0! to 4!
+// Introduced to avoid 'magic number's clang-tidy complaints about
+constexpr int fact_array_size = 5;
+constexpr std::array<float, fact_array_size> fact = {1., 1., 2., 6., 24.};
+
+// Iterates over first `n_batch_dims` and applies `f` to each element.
+// TODO: this implemtation is not efficient, it simply iteratates over
+// batch elements in a serial for-loop.
+template <typename func_t>
+void batch_apply(at::ArrayRef<Tensor> tensors, int64_t n_batch_dims, const func_t& f) {
+  // no-op if batch is empty
+  if (n_batch_dims <= 0) {
+    return;
+  }
+
+  std::vector<Tensor> tensors_view;
+  for (auto& tensor : tensors) {
+    std::vector<int64_t> tensor_view_sizes = {-1};
+    for (auto d = n_batch_dims; d < tensor.dim(); ++d) {
+      tensor_view_sizes.push_back(tensor.size(d));
+    }
+    auto tensor_view = tensor.view(tensor_view_sizes);
+    tensors_view.push_back(tensor_view);
+  }
+
+  for (int64_t input_idx = 0; input_idx < tensors_view[0].size(0);
+    ++input_idx) {
+    std::vector<Tensor> input_selected;
+    for (auto& tensor_view : tensors_view) {
+      auto curr_input = tensor_view.select(0, input_idx);
+      input_selected.push_back(curr_input);
+    }
+    f(input_selected);
+  }
+}
+
+Tensor matrix_power(const Tensor& matrices, const Tensor& powers) {
+  if (matrices.dim() > 2) {
+    auto res = at::empty(matrices.sizes(), matrices.options());
+
+    batch_apply(
+      {res, matrices, powers},
+      matrices.dim() - 2,
+      [](auto tensors) {
+        auto& out = tensors[0];
+        auto& in = tensors[1];
+        int64_t n = tensors[2].item().template to<int64_t>();
+        out.copy_(at::matrix_power(in, n));
+      }
+    );
+
+    return res;
+  }
+  else {
+    int64_t n = powers.item().template to<int64_t>();
+    return at::matrix_power(matrices, n);
+  }
+}
+
+Tensor operator_1_norm(const Tensor& tensor) {
+  return std::get<0>(tensor.abs().sum(-2).max(-1));
+}
+
+Tensor compute_T1(const Tensor& A) {
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  return I + A;
+}
+
+Tensor compute_T2(const Tensor& A) {
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  const auto A2 = at::matmul(A, A);
+  return I + A + A2 / fact[2];
+}
+
+Tensor compute_T4(const Tensor& A) {
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  const auto A2 = at::matmul(A, A);
+  return I + A + at::matmul(A2, I / fact[2] + A / fact[3] + A2 / fact[4]);
```

Comment:
Same as above.

Reply:
No, here we cannot. The factorial is shifted one power to the right.